src/auth-module.ts

```diff
@@ -210,4 +259,163 @@ export class AuthModule implements NestModule, OnModuleInit {
 			],
 		};
 	}
+
+	/**
+	 * Static factory method to create and configure the AuthModule asynchronously.
+	 * @param options - Async configuration options for the module
+	 */
+	static forRootAsync(options: AuthModuleAsyncOptions): {
+		global: boolean;
+		module: typeof AuthModule;
+		imports?: ModuleMetadata["imports"];
+		providers: Provider[];
+		exports: (Provider | typeof AuthService)[];
+	} {
+		const asyncProviders = AuthModule.createAsyncProviders(options);
+
+		return {
+			global: true,
+			module: AuthModule,
+			imports: options.imports || [],
+			providers: [...asyncProviders, AuthService],
+			exports: [
+				{
+					provide: AUTH_INSTANCE_KEY,
+					useExisting: AUTH_INSTANCE_KEY,
+				},
+				{
+					provide: AUTH_MODULE_OPTIONS_KEY,
+					useExisting: AUTH_MODULE_OPTIONS_KEY,
+				},
+				AuthService,
+			],
+		};
+	}
+
+	private static createAsyncProviders(
+		options: AuthModuleAsyncOptions,
+	): Provider[] {
+		if (options.useFactory) {
+			return [
+				{
+					provide: AUTH_INSTANCE_KEY,
+					useFactory: async (...args: unknown[]) => {
+						const result = await options.useFactory?.(...args);
+						const auth = result.auth;
+
+						// Initialize hooks with an empty object if undefined
+						auth.options.hooks = {
+							...auth.options.hooks,
+						};
```

Comment:
According to the typing, the correct one would be `auth.hook`.

```suggestion
						// Initialize hooks with an empty object if undefined
						auth.hooks = {
							...auth.hooks,
						};
```

I had to do this workaround in my `useFactory` to make it work.

```ts
auth: {
  // @ts-expect-error - options is not typed
  options: {
    hooks: {}
  }
}
```

If I didn't do this the error would be:
```
TypeError: Cannot read properties of undefined (reading 'hooks')
```

Reply:
Thanks @mwolf1989 - this one I'm having a little trouble with.

If I try to change to `auth.hooks` I am getting a TS compilation error

`Property 'hooks' does not exist on type 'Auth'.ts(2339)`

Looking at the `better-auth` types, I see:

```ts
type Auth = {
    handler: (request: Request) => Promise<Response>;
    api: FilterActions<ReturnType<typeof router>["endpoints"]>;
    options: BetterAuthOptions;
    $ERROR_CODES: typeof BASE_ERROR_CODES;
    $context: Promise<AuthContext>;
};
```

And then I see the expected `hooks` property under `options` - what am I missing here?

I.e, if I try supplying the `hooks` property to a real `betterAuth` instance in the options, I see no issues

```ts
    AuthModule.forRootAsync({
      useFactory: (database: NodePgDatabase) => ({
        auth: betterAuth({
          database: drizzleAdapter(database, {
            provider: 'pg',
          }),
          hooks: {},
          emailAndPassword: {
            enabled: true,
          },
          trustedOrigins: ['http://localhost:3001'],
        }),
      }),
      inject: [DATABASE_CONNECTION],
    }),
```

Appreciate your help on this.

Reply:
Thanks @ThallesP - so if I'm understanding correctly, this issue comes down to compatibility around the `Auth` type provided by `better-auth` in different versions.

I've gone ahead and updated the PR to replace the `Auth` type with `any`, as well as updating `biome.json` to allow for it. Hopefully, we can come back to this in the future.

In terms of the `auth.hooks` issue, I *think* this check should allow for support on both variants, but please correct me if I'm misunderstanding the issue or you have an alternative suggestion.

```ts
// Initialize hooks with an empty object if undefined
// Handle both auth.hooks and auth.options.hooks for compatibility
if (auth.hooks !== undefined) {
	auth.hooks = { ...auth.hooks };
} else if (auth.options) {
	auth.options.hooks = { ...auth.options.hooks };
}
```

CC @mateus4k

Reply:
Edit: Added inline biome skip instead of changing `biome.json` directly.

Reply:
> Thanks @ThallesP - so if I'm understanding correctly, this issue comes down to compatibility around the `Auth` type provided by `better-auth` in different versions.
> 
> I've gone ahead and updated the PR to replace the `Auth` type with `any`, as well as updating `biome.json` to allow for it. Hopefully, we can come back to this in the future.
> 
> In terms of the `auth.hooks` issue, I _think_ this check should allow for support on both variants, but please correct me if I'm misunderstanding the issue or you have an alternative suggestion.
> 
> ```ts
> // Initialize hooks with an empty object if undefined
> // Handle both auth.hooks and auth.options.hooks for compatibility
> if (auth.hooks !== undefined) {
> 	auth.hooks = { ...auth.hooks };
> } else if (auth.options) {
> 	auth.options.hooks = { ...auth.options.hooks };
> }
> ```
> 
> CC @mateus4k

great, feel free to drop that `auth.hooks` check, i'm pretty sure it's incorrect and from the start better-auth always used auth.options.hook.

Reply:
Thanks @ThallesP - just backed it out.